sw/airborne/modules/flight_plan_guided/flight_plan_guided.c

```diff
@@ -0,0 +1,427 @@
+ *   will not be necessary anymore
+ */
+
+/**
+ * RotateToHeading_ATT: Rotate to heading in ATT mode
+ * @param[in] heading, to turn to in a guided flight [rad] (clockwise is positive)
+ * @param[in] trim_phi, during turn, apply a small trim to the roll if sideways drift is noticeable [rad]
+ * @param[in] trim_theta, during turn, apply a small trim to the pitch if forward/backwards drift is noticeable
+ * @param[in] in_flight, bool necessary for stabilization loop to check if the MAV is flying
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool RotateToHeading_ATT(float heading, float trim_phi, float trim_theta, bool in_flight)
+{
+  struct Int32Eulers cmd;
+
+  if (guidance_h.mode == GUIDANCE_H_MODE_ATTITUDE) {
+    cmd.phi = ANGLE_BFP_OF_REAL(trim_phi); //trim?
+    cmd.theta = ANGLE_BFP_OF_REAL(trim_theta);
+    cmd.psi = ANGLE_BFP_OF_REAL(heading);
+
+    stabilization_attitude_set_rpy_setpoint_i(&cmd);
+    stabilization_attitude_run(in_flight);
+  }
+  return false;
+}
+
+/**
+ * ResetAngles_ATT: Reset all angles while in ATT mode
+ * @param[in] current_heading, indicated the current heading of the MAV [rad] (clockwise is positive)
+ * @param[in] in_flight, bool necessary for stabilization loop to check if the MAV is flying
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool ResetAngles_ATT(float current_heading, bool in_flight)
+{
+  struct Int32Eulers cmd;
+  if (guidance_h.mode == GUIDANCE_H_MODE_ATTITUDE) {
+    cmd.phi = ANGLE_BFP_OF_REAL(0.0f);
+    cmd.theta = ANGLE_BFP_OF_REAL(0.0f);
+    cmd.psi = ANGLE_BFP_OF_REAL(current_heading);
+
+    stabilization_attitude_set_rpy_setpoint_i(&cmd);
+    stabilization_attitude_run(in_flight);
+  }
+  return false;
+}
+
+/********************************************//**
+ *  Blocking conditions
+ ***********************************************/
+
+/**
+ * WaitUntilAltitude: Blocks while waiting until a certain altitude is reached
+ * @param[in] altitude in [m] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilAltitude(float altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (stateGetPositionEnu_f()->z < -altitude) { return true; }
+
+  return false;
+}
+
+/**
+ * WaitUntilSpeedOrAltitude: Blocks while waiting until a certain altitude or speed is reached
+ * @param[in] fail_altitude, wait for altitude in [m] (positive z-axis is up)
+ * @param[in] speed, speed in [m/s] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilSpeedOrAltitude(float speed, float fail_altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (stateGetPositionEnu_f()->z > -fail_altitude) { return false; }
+  if (stateGetSpeedEnu_f()->z < -speed) { return true; }
```

Comment:
Same strange negation here. Also why fail_altitude and not just altitude like you have speed.

if (stateGetPositionEnu_f()->z < altitude) { return true; }
if (stateGetSpeedEnu_f()->z < speed) { return true; }